dango/client/src/signer.rs

```diff
@@ -55,37 +70,70 @@ where
 }
 
 impl SingleSigner<Undefined<u32>> {
-    pub fn new(username: &str, address: Addr, sk: SigningKey) -> anyhow::Result<Self> {
+    pub fn new(
+        username: &str,
+        address: Addr,
+        sk: SigningKey,
+        credential_type: CredentialType,
+    ) -> anyhow::Result<Self> {
         let username = Username::from_str(username)?;
 
+        let (key, key_hash) = match credential_type {
+            CredentialType::Secp256k1 => (
+                Key::Secp256k1(ByteArray::from_inner(sk.public_key())),
+                sk.public_key().hash256(),
+            ),
+            CredentialType::Ethereum => {
+                let addr =
+                    Addr::from_inner(keccak256(&sk.extended_public_key()[1..])[12..].try_into()?);
```

Comment:
We have `hyperlane_testing::eth_utils::derive_address` function for this. You can move this whole module to `grug_types` (put under and optional `ethereum` flag) if you need to use it outside of Hyperlane context.

Reply:
Done in [abc4b16](https://github.com/left-curve/left-curve/pull/974/commits/abc4b168e49cbfe050e5436b5b0be309ae4832c8)